### PR TITLE
[Executors] Ensure we treat DA older than 5.9 always as DefaultActor

### PR DESCRIPTION
**Pull request by ktoso**
This is because cross module references to a RESILIENT actor would have short circuited and returned false for such actor, while an old compiler does not have code to handle the "NON default distributed actor" paths, so it would result in compile issues in HopToActor optimization

### CODE REVIEW

File: lib/Sema/TypeCheckConcurrency.cpp [hunk 1]
```diff
@@ -182,6 +182,27 @@ bool IsDefaultActorRequest::evaluate(
...
+    auto actorAvailability = TypeChecker::overApproximateAvailabilityAtLocation(
+        classDecl->getStartLoc(),
+        classDecl);
+
+    if (!actorAvailability.isContainedIn(customExecutorAvailability)) {
+      // Any 'distributed actor' declared with availability lower than the
+      // introduction of custom executors for distributed actors, must be treated as default actor,
+      // even if it were to declared the unowned executor property, as older compilers
+      // do not have the the logic to handle that case.
+      return true;
```
Comment (ktoso) [inline lib/Sema/TypeCheckConcurrency.cpp:202]: This is the core of the change; we ensure we treat distributed actors which "could not have been non-default" always as default.

File: test/Distributed/Runtime/distributed_actor_custom_executor_availability.swift [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+      }
+
+      tests.test("5.9 actor, no availability executor property => custom executor") {
+        try! await FiveNineActor_NothingExecutor(actorSystem: system).test(x: 42)
+      }
+
+      tests.test("5.7 actor, 5.9 executor property => no custom executor") {
+        expectCrashLater(withMessage: "Fatal error: Incorrect actor executor assumption; Expected 'MainActor' executor.")
+        try! await FiveSevenActor_FiveNineExecutor(actorSystem: system).test(x: 42)
+      }
```
Comment (ktoso) [inline test/Distributed/Runtime/distributed_actor_custom_executor_availability.swift:100]: This case I was unsure about, but I think this way to handle it makes sense; the entire actor has to be 5.9 to be treated like being able to have a custom executor, not just the property.

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
Going to debug the linux issue 👀

**Comment by ktoso** [comment]
Fixed non apple platform issues:
- silly extra space issue in test: [`a6ddf44` (#64800)](https://github.com/apple/swift/pull/64800/commits/a6ddf4460d92e32ae006b629c906ccc45b4d5201)
- looking for macos availability should have no impact on linux, so adjusted the test for this: [`3b4ac6c` (#64800)](https://github.com/apple/swift/pull/64800/commits/3b4ac6c821ed4bf66292815cfe3719358905710d)

**Comment by ktoso** [comment]
@swift-ci please smoke test

File: include/swift/AST/Availability.h [hunk 3]
```diff
@@ -329,6 +329,11 @@ class AvailabilityContext {
   bool isAvailableAsSPI() const {
     return SPI && *SPI;
   }
+
+  /// Returns a representation of this range as a string for debugging purposes.
+  std::string getAsString() const {
+    return "AvailabilityContext(" + OSVersion.getAsString() + (isAvailableAsSPI() ? ", spi" : "") + ")";
```
Comment (tshortli) [inline include/swift/AST/Availability.h:335]: Minor: run `git-clang-format`

**Review by tshortli — APPROVED** [review]

File: include/swift/AST/Availability.h (hunk 3 above)
Comment (ktoso) [reply include/swift/AST/Availability.h:335]: Thanks, I'll follow up separately if that's fine? That'd just be a linebreak here AFAICS.

```diff
-    return "AvailabilityContext(" + OSVersion.getAsString() + (isAvailableAsSPI() ? ", spi" : "") + ")";
+    return "AvailabilityContext(" + OSVersion.getAsString() +
+           (isAvailableAsSPI() ? ", spi" : "") + ")";
```

// The problem with clang-format is that it is not being applied consistently in this project, so applying it adds all kinds of noise and "this commit does not look like the rest of the file or method at all now..."

Comment (tshortli) [reply include/swift/AST/Availability.h:335]: Follow up is fine with me! I find the `clang-format` output is pretty consistent with the style of most of the files I have worked in in the project and it is documented in https://github.com/apple/swift/blob/main/docs/HowToGuides/FirstPullRequest.md as the primary way of enforcing the project code style. In particular, long lines stick out a lot - I think the majority of the codebase adheres to the 80 char limit

Comment (ktoso) [reply include/swift/AST/Availability.h:335]: I'll follow up about my specific changes, but the reality is that files are not well formatted: 

```
 22 files changed, 65 insertions(+), 65 deletions(-)
```

if running clang-format on files this PR touched, out of that my changes would be...

```
 5 files changed, 22 insertions(+), 15 deletions(-)
```

but yeah, I'll add a git hook to git-clang-format my commits and I'll do the promised follow up now as well

**Pull request merged**